File: go/day15part2.go
package main
/*
# PROBLEM:
#    The "Elf Game" has the following rules:
#       * There is a starting set of numbers
#       * At each turn AFTER the initial set of numbers
#         has been spoken, the player must either state
#         0 - the previous number has never been seen before
#         n - The previous number came up n turns ago.
#         (strictly speaking the difference between the last 2
#          utterences.)
#
#  Given an input starting set, determine the 30000000th  number spoken.
*/

import(
  "fmt"
  "log"
  "time"
)

/*
------------------------------------------------------------------------------
Initialise the game by creating the "last seen" map
*/
func startElfGame(startNumbers []int) (int, map[int]int, map[int]int) {
  lastMap := map[int]int{}
  prevMap := map[int]int{}
  turn := 1
  for i := range(startNumbers) {
    lastMap[startNumbers[i]] = turn
    prevMap[startNumbers[i]] = 0
    turn++
  }
  return turn,lastMap,prevMap
}


/*
------------------------------------------------------------------------------
*/
func playElfGame(startNumbers []int, iterations int) int {
  i,lastMap,prevMap := startElfGame(startNumbers)
  fmt.Printf("Game initialised as: %#v, %#v and next turn %d\n",lastMap,prevMap,i)
  lastSpoken := startNumbers[len(startNumbers)-1]
  for i <= iterations {
    nextSpoken := 0 //A handy default
    //Game Rules:
    // Examine the answer from the last iteration. Has it been spoken before?
    lastSeen,lFound := lastMap[lastSpoken]
    prevSeen,pFound := prevMap[lastSpoken]
    if (lFound && !pFound) || (!lFound && pFound) { //Catch a data-error
      log.Fatal("Data error - no match between last/previous found maps for v=", lastSpoken)
    } else if !lFound && !pFound {   //We've never seen this before, answer is Zero (as above)
      lastMap[lastSpoken] = i-1 //Initialise history for this
      prevMap[lastSpoken] = 0
    } else {
      if prevSeen == 0 {  //We've seen this once before. Special case, was it on last turn?
        if lastSeen == i-1 { //Yes - answer Zero
          nextSpoken = 0
        } else { //No - Answer with diff
          nextSpoken = (i-1) - lastSeen
        }
      } else { //We've seen the number X times before...
        nextSpoken = lastSeen - prevSeen
      }
      //Update stored values if we ever saw this before:
      t:=lastMap[nextSpoken]
      prevMap[nextSpoken] = t
      lastMap[nextSpoken] = i
    }
    //Get ready for the next turn...
    i++
    lastSpoken = nextSpoken
  }
  return lastSpoken
}

/*
------------------------------------------------------------------------------
*/
func runHarness(startNumbers []int, iterations int) int64 {

  start := time.Now()
  answer := playElfGame(startNumbers,iterations)
  t := time.Now()
  elapsed := t.Sub(start)
  turnRate := int64(iterations*1000000) / elapsed.Microseconds()

  fmt.Printf("Answer after %-8d iterations: %-5d (took %.6f seconds; %-8d turns/second)\n",
    iterations,
    answer,
    (elapsed.Seconds() * 1e6) /1e6,
    turnRate,
  )
  return turnRate
}

/*
------------------------------------------------------------------------------
------------------------------------------------------------------------------
------------------------------------------------------------------------------
*/
func main() {

  //Test Data i=2020,a=438. i=30000000,a=18
  //startNumbers := []int{3,2,1}
  //Test Data i=2020, a=436. i=30000000, a=175594
  //startNumbers := []int{ 0,3,6 }
  //Actual puzzle input:
  startNumbers := []int{ 0,5,4,1,10,14,7 }

  turnRate := runHarness(startNumbers,2020)
  fmt.Printf("Predicted execution time for 202020 turns: %.4f seconds\n",
    float64(202020) / float64(turnRate),
  )

  turnRate = runHarness(startNumbers,202020)
  fmt.Printf("Predicted execution time for 30000000 turns: %.4f seconds\n",
    float64(30000000) / float64(turnRate),
  )

  turnRate = runHarness(startNumbers,30000000)




}

File: python/day15part2.py
#!/usr/bin/python3
#
# My working for Day 15 Part 2
#
# PROBLEM:
#    The "Elf Game" has the following rules:
#       * There is a starting set of numbers
#       * At each turn AFTER the initial set of numbers
#         has been spoken, the player must either state
#         0 - the previous number has never been seen before
#         n - The previous number came up n turns ago.
#         (strictly speaking the difference between the last 2
#          utterences.)
#
#  Given an input starting set, determine the 30000000th  number spoken.
#
# DISCUSSION: This is a trivial extension of the first part. So we'll just
#             time the answers to get an idea of how long it'll take to
#             do it, but I suspect the answer is BRUTE FORCE.
import time

#How long should we play for?
bigTarget = 30000000


#Test Data. Answer = 1
startNumbers = [ 1,3,2 ]

#Test, a= 10
startNumbers = [ 2,1,3 ]
#Test, a=27
startNumbers = [1,2,3]

#Test Data i=2020,a=438. i=30000000,a=18
startNumbers = [3,2,1]

#Test Data i=2020, a=436. i=30000000, a=175594
startNumbers = [ 0,3,6 ]

#Actual puzzle input:
startNumbers = [ 0,5,4,1,10,14,7 ]

#Here is a Glob to hold our number data.
nData = {}

###################################################################
#Setup the game:
def init_nData(refNumbers) :
    global nData
    i=1
    while i <= len(refNumbers) :
        nData[refNumbers[i-1]]={ 'l' : i, 'p' : 0 }
        i +=1
    return i
###################################################################
#Play the game:
def play_game(initNumbers,targetTurn) :
    global nData
    i = init_nData(startNumbers)
    pNum = startNumbers[len(startNumbers)-1]
    tStart = time.time()
    while i <= targetTurn :
        #RULES:
        #  IF pNum was FIRST SPOKEN last turn, ANS=0
        #  ELSE: ANS = diff(last turn, last spoken)
        nextNumber = 0
        if pNum not in nData :
            #Satifies criteria "never previously spoken"
            pass
        elif nData[pNum]['p'] == 0 :
            #Number has only been spoken once before.
            if nData[pNum]['l'] <= i-1 :
                #...but it was on a previous turn
                nextNumber = (i-1) - nData[pNum]['l']
        else :
            #Number has definitely been spoken before.
            nextNumber = nData[pNum]['l'] - nData[pNum]['p']
        pNum=nextNumber
        if nextNumber not in nData :
            nData[nextNumber] = {'l' : i, 'p' : 0 }
        else :
            newPrev = nData[nextNumber]['l']
            nData[nextNumber] = { 'l' : i, 'p' : newPrev}
        i += 1

    print(f"Finishing turn {i-1}, last number spoken was {nextNumber}")
    tEnd = time.time()
    tDiff = tEnd - tStart
    tRate = targetTurn / tDiff
    print(f"(elapsed time: {tDiff:3.3f}, turns/sec: {tRate:1.2f})")
    return nextNumber, tRate

###################################################################
###################################################################
if __name__ == '__main__':
    print(startNumbers)
    ans,rate = play_game(startNumbers,2020)
    estTimeToBigTarget = bigTarget / rate
    print(f"Estimated time to finish {bigTarget} rounds: {estTimeToBigTarget:1.2f} Seconds")
    print("-"*80)
    nData={} #See if you can spot the last code edit to solve the "IT'S NOT WORKING ON BIG ITERATIONS!" arrghg...
    ans,rate = play_game(startNumbers,bigTarget)
